### DISCUSSION TITLE
Reading Data from Wifi RS485 connector

### DISCUSSION THREAD

**Discussion by valdiressa** [macxq/foxess-ha#49] [Show and tell]
Hi everyone this is my first time ever posto on Github. I couldn't help posting because I'm been looking on the internet for a solution similar to the one provided by this integration but this is ONLY forum i found anything even remotely related to the extraction of data from a Foxess inverter.

So i've been playing with similar but came to a block in my investigations that i cant see to find a way out. I'm been "sniffing" data that the inverter sends to the wifi card directly from the RS485 connector. The code is ofcourse coded in a way but i cant seem to make sense of find a way to identify each of the values being sent from the port.

There are mostly 3 burst of data that the inverter sends from the connector (there are more but 3 are mostly repetitive) these are something like:

b'0303fd3d0d9e26fdaf99a39f8fffff99a39f9dffff99a39f99fffffffd579b579b95ffffffffffffffffffffffffff3daf380a62c4f8'
b'0303fb3d0d9eaefed3fe4de9dfd22ddeadfea7d9e1d2dfdecdfea5d9e1d2dfde99fea5d9e1d2dfc64ffea5ffffe5c3feb1ffffffffffffffffffffffffffffffa9ffa9ffa3ffc9ffff3393fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffb93836262fc'
b'0303f33d0d9e36fed57b9d9f9d9f9f93939b579b95999f9d9b5f7b9f9f9d5bc73162fe'

it sends mostly one of 35 byte and 30s later 54 bytes and 30s later 164 bytes. Does this all the time. I believe that the esp32 inside the wifi card computes this data and sends it over the internet to the Foxess cloud.

I opened the wifi card and soldered some wires to have access to the RS485 datalines.

![IMG_20220127_161453](https://user-images.githubusercontent.com/14052501/153418509-4585ebd9-74ef-432b-8ea2-2fb7d3e6040f.jpg)
![IMG_20220127_154911](https://user-images.githubusercontent.com/14052501/153418521-38562183-af31-4f28-9a05-13bb18760d1d.jpg)

I'm tried to make sense out of the data and even made some advances in identifying changes in the stream of data when i make changes to the inverter like shutdown a PV1 input or PV2 input...so data is here and we could extract directly from the inverter just need to understand what calculations are needed to read the attached.

Hope we can start a conversation about this.

Regards,

**Comment by williamjeccles** [comment]
Really neat idea, love it. Can't contribute at this time as it's out of my field of expertise but I know some people who I'll be sure to run the problem by.

**Comment by macxq** [comment]
@valdiressa  Interesting project, thx for sharing it with us :) 

1. Can do share more detail to what device the the soldered cables are connected ? is it Arduino or just a PC and how you are receiving them?
2. Why you decided to open the wifi dongle ? Do you have any informations using what protocol this dongle is communicating with the inverter?
3. Why didn't you decided to connect to dedicated  **F** `communication port ` 
![Screenshot 2022-02-10 at 21 15 26](https://user-images.githubusercontent.com/2965092/153489887-bcc6b06e-efba-4162-ba1a-7e62e1d04da4.png)
there is a clear description of wiring:
![Screenshot 2022-02-10 at 21 15 42](https://user-images.githubusercontent.com/2965092/153489990-d0428f67-1aff-4f81-b3de-f8d478a14219.png)
4. How do you want to use this data from inverter? 
5. If you want to push it to Home Assistant. I'm recommending [arduino-modbus-rtu-tcp-gateway](https://github.com/budulinek/arduino-modbus-rtu-tcp-gateway) project  and receiving them over this [integration](https://www.home-assistant.io/integrations/modbus/)  
6. [There are already some out-of-the-box cheap solutions](https://pl.aliexpress.com/item/32916128353.html?spm=a2g0o.productlist.0.0.1b88495bKvU9YT&algo_pvid=ab612b87-da8f-4042-91fd-3d57901c5a53&algo_exp_id=ab612b87-da8f-4042-91fd-3d57901c5a53-2&pdp_ext_f=%7B%22sku_id%22%3A%2210000004043121727%22%7D&pdp_pi=-1%3B69.93%3B-1%3B-1%40salePrice%3BPLN%3Bsearch-mainSearch)

Thank you for blazing the trail, who know if we will not be pushed to go this route.

**Reply by valdiressa** [reply]
> @valdiressa Interesting project, thx for sharing it with us :)

> 1. Can do share more detail to what device the the soldered cables are connected ? is it Arduino or just a PC and how you are receiving them?
I used a Rapsberry PI with a dongle USB to RS485 and captured the data from python and Nodered.
> 2. Why you decided to open the wifi dongle ? Do you have any informations using what protocol this dongle is communicating with the inverter?
I don't have any documentation besides that API pdf document that talks about the MQTT from foxess cloud setup that didn't work for me. Why I used the wifi dongle is linked to the question below please check. Plus remember ALL data that goes to Foxess Cloud comes from this port and if we "sniff" this data we have all of it without needing know the protocol that is required "ask the inverter for it". We just need to know the protocol to read this data comes out of the inverter every 30s or so.
> 3. Why didn't you decided to connect to dedicated  **F** `communication port `
So this was my very first option and with limited success. I have a Foxess T25 Ongrid inverter very similar to the picture setup you shared but quite larger inverter. This inverter communicates with a Smart Meter from CHINT ref - DTSU666 and the reason for this communication is to achieve Zero Export to the grid. My country does not allow grid exports.
These datalines use RS485 RTU protocol and on this protocol there is only one Master on the line. In this case the master is the Inverter and being the master floods the line with the same QUERY ALL the time....this line is always busy.....and the meter is always being asked the same thing...."Reading of the Power being exported".
The inverter takes this information and shows on the screen plus it uses it to clamp down on the power production from the panels. Always trying to achieve zero power export.
Still I manage to sniff and actually query the meter from this line (with many timeouts) on information like the Voltages and Currents from each of the phases (my house is 3 phase 220V AC power from Grid). 
> 4. How do you want to use this data from inverter?
So my setup is very complex. I have OnGrid and OffGrid inverters plus a High Voltage (220V DC - 202ah - 66series 3,2v - custom build) battery pack and I need to manage the charging, discharging plus do some heavier load management. The panels also need to move from the offgrid solar controller to the ongrid inverter after the battery packs are full. So I need data to control all these different states of the system.   
> 5. If you want to push it to Home Assistant. I'm recommending [arduino-modbus-rtu-tcp-gateway]
Thank you for this will checkout for future. I'm currently using NodeRed for most data processing and control. For data input and specially if looking at RS485 its much easier to deal with NodeRed (transforming into MQTT data) directly because you can do it all in the software and model your data to be consumed directly to any other client you have like influxdb, Grafana, etc. 
So I could use arduino as a gateway I just don't see benefit yet :)
> 6. [There are already some out-of-the-box cheap solutions]
Thanks for this will look in to it. It does look like a good solution if the "Gateway" is far from the inverter.

> Thank you for blazing the trail, who know if we will not be pushed to go this route.

Thank you. I think that reading directly from the inverter is definitely the best and most secure way to go. Lets hope I can find more info on the protocols to be able to decode this. Having to check Hex data strings and try to decode this is not simple..... 

Regards,
Valdir

**Comment by valdiressa** [comment]
Also before i'm asked :). After looking at the Pin 3 and 4 and seen that it was very busy I shifted to pins 1 and 2 hoping that it was not but just to find that it was the exact same.

**Comment by Jmcullimore** [comment]
That looks like an Espressif chip, which would hold up why I have one on my network, can you post a clearer photo of the back of the shiny metal casing that's near the SMT connector, does it have a chip number or something?

**Reply by Jmcullimore** [reply]
Just took mine apart, it's a wroom-02u esp8266!

**Comment by valdiressa** [comment]
Hi all, spent a few days checking and trying to understand data being sent out but still no luck. Unfornately I am also making some major changes to the system that I probably not need the FOxEss inverter in my setup anymore. If someone else works on this issue let me know I'll help out as much as i can with the data I've already acquired.

**Reply by StealthChesnut** [reply]
Hi @valdiressa, did you (or @Jmcullimore) get a photo of the back of the wifi board as well please?

**Reply by valdiressa** [reply]
![image](https://user-images.githubusercontent.com/14052501/161023942-9a00b562-d401-4e00-abc8-b993564a12ea.png)
Is this what you are looking for!?

**Reply by StealthChesnut** [reply]
Yes, many thanks!

I've been able to get some data from the inverter, but the addresses don't match up with any specs or definitions. FoxESS don't reply to me asking them for documentation. Do you know what any of the addresses / data fields are?

**Reply by valdiressa** [reply]
Hi there. Well thats the biggest issue. Lack of documentation.

So if you are using this port's RS485 you actually don't need to query anything because the wireless board already queries EVERYTHING that the FOXess cloud tracks. So the answers (data) for our possible queries are already passing on the wire.

The issue is exactly that "what is what"?

Funny thing is that the same inverter talking to the Meter I did manage to extract data and decode it....but again this is the inverter talking the Meter language and addresses.....

The inverter does not use the same protocol on the wireless port.

**Reply by MattBatesWasNotAvailable** [reply]
Hello,

I'm also trying to figure out how to get modbus data out of the inverter, the documentation says its possible at least over the RS485 port but it also sort of says it should work over ethernet or wifi as well.

Has anyone successfully made a modbus connection to the inverter either via wifi or ethernet? I've tried using a few modbus debugging tools but haven't yet been able to make a connection.

My last resort idea is to go from RS485 to usb into a low powered device (like a raspberry pi) to use a network proxy.

**Reply by StealthChesnut** [reply]
The route I've used is RS485-to-USB, have been getting data reliably for a couple of weeks now. I've been trying to figure out a few more data fields but will share my findings. I'm not sure how much further I can get without proper info from Fox.

**Reply by StealthChesnut** [reply]
I have shared my findings (and others have also contributed) here: https://github.com/StealthChesnut/HA-FoxESS-Modbus

Hope it helps you in some way!

**Comment by Daibutt** [comment]
@MattBatesWasNotAvailable 
Quite a bit of dicussion here https://www.facebook.com/groups/foxessownersgroup/permalink/1549648805417296/
I'd say it looks promising but not fully there yet.

**Reply by valdiressa** [reply]
I checked this link and its very promissing and one of the guys there actually read data from the RS485 port (NOT the WIFI RS485 port) so its possible.
[FoxESS Modbus On Grid solar inverter Modbus Protocol.pdf](https://github.com/macxq/foxess-ha/files/8552500/FoxESS.Modbus.On.Grid.solar.inverter.Modbus.Protocol.pdf)

It looks like he had luck on the registers from 10000 to10027 and 11000 to 11095. I didn't try myself and unfortunately i actually disconnected the Foxess Inverter at least for a while as i'm testing other equipment setup at the moment.

But there is a problem with reading this Rs485 port "if" you are using a SmartMeter like i am with configured Zero Export. This port is "bombarded" with requests from the Inverter to Chint Meter for updates on the "PowerExport" variable on the meter. It is an immensely busy port.

So if you have the same setup that i have/had that will be a problem since if you add another query on the same line for registers in the Inverter that will have lots of timeouts due to the constant reads from the Inverter - Meter comms.

That is why my best bet was the RS485 Wifi port.....this is basically always sending data and thats the exact data we are all looking for already.

**Comment by straga** [comment]
I found more info about protocol.
https://github.com/nigelspowage/foxesscloud_protocol/issues/1
https://github.com/nigelspowage/foxesscloud_protocol/wiki

**Reply by valdiressa** [reply]
This is very interesting but it does require to check on TCP side and possibly validate the link between the data and the registers (Rs485). From some of the comments in the forum it looks like the port does not need to have the wifi board connected to send data out in the already seen timed bursts of data. I didn't verify this but if that's the case it might be easier to use this port and what we already know about the RS485 protocol to query the inverter... seeing that the port is not as busy as the other one (of course this is my case that i use a smartmeter).

**Reply by straga** [reply]
I am sending an email to Fox, to ask about protocol info.

**Reply by So5nA** [reply]
Modbus works there, but it doesn't need any queries - it sends frames every 30 seconds (real time info), so there probably can't be another device on this rs485 line - there will be too much noise.
Manager version 3.10 or higher is required on the inverter.

**Reply by straga** [reply]
> Modbus works there, but it doesn't need any queries - it sends frames every 30 seconds (real time info), so there probably can't be another device on this rs485 line - there will be too much noise. Manager version 3.10 or higher is required on the inverter.

Foxess Sxxxx series? When connecting the meter too? Inverter ask meter and send info frame every 30 sec?

**Reply by So5nA** [reply]
Nope, I've T series.
Don't have meter. My tests show that the inverter is not listening on 485 (in my case on pin 1&2). It sends a frame every 30 seconds.
![cone](https://user-images.githubusercontent.com/100235481/165290786-5c138cca-8ea6-45d1-9d03-7365a2604d33.jpg)
The frame is a little changed compared to the modbus.
The start of the frame is 2 bytes (7E7E), and the end is E7E7.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">




</head>

<body link="#0563C1" vlink="#954F72">



Frame   header | Function | Timestamp | Data Length | Data | MODBUS CRC16 | Frame end
-- | -- | -- | -- | -- | -- | --
2 bytes | 1 byte | 4 bytes | 2 bytes | x bytes | 2 bytes | 2 bytes
7E 7E | 01 | 62 67 DC 48 | 00 28 | 40 bytes | CRC16 | E7 E7
7E 7E | 02 | 62 67 DC 48 | 00 96 | 150   bytes | CRC16 | E7 E7
7E 7E | 06 | 62 67 DC 48 | 00 15 | 21 bytes | CRC16 | E7 E7



</body>

</html>

**Reply by valdiressa** [reply]
Hi @So5nA i also have the T25 series and I'm actually using that specific port with 3 and 4 to the METER. Also tried to use 1 and 2 just to find its the same lines as 3 and 4. Different from your setup I do have a Meter there is ALOT of traffic in this line. Inverter being the master does queries much faster than 30s (its actually in the ms) to the Meter to get the Power Export value from the Meter and regulate it self for ZeroExport. On this line i could monitor the communication from the Inverter to the Meter and the Inverter actually used a standard the protocol used to query the CHINT meter. I didn't however try to query the Inverter. But I do not expect to be very successful due to constant traffic on the line. Many of my queries to the Meter (on the same line) had timeouts and could only get replies every 25 secs or so after multiple failures.

Example of comms:
b'
3843 - Start Frame
01   - Address Slave/Meter
03   - Function Code
2012 - Power data ModBus equivalent to register 8210 in the Meter
0002 - NUmber of Registers to read
6f ce 01 03 - Raw data 4 bytes
04 
c5 - Error Check
0a' - end Frame

**Reply by assembly12** [reply]
@So5nA 
do you have RS485/Modbus connection working on your T-Series?
How did you do it?

**Comment by straga** [comment]
I can test next week Sxxxx, I will try - a parallel connection to the Wifi adapter like in the first post.

**Comment by straga** [comment]
I have a question about Export Limiter. Does anybody use it? It is possible to get 0w to export. 
I get for test S2500, connect Meter. And that not work like Solax with EXport Limiter.
Always export or import any value.

**Reply by valdiressa** [reply]
Hi there. I used it. Its possible to get "near 0 to 0" I say that because of the reaction time. How it works is that an independent meter is placed at the Grid input of the site/house and it will monitor the power being exported to the grid and will feedback this info (via RS485) all the time to the inverter. The inverter will throttle down the energy production of the Solar Cells until it reaches 0 and will keep doing that all the time. If the loads increase it will try to increase the production to match if the load drops it will check if its putting out to the grid and cut it down reactively. Because its a reactive process its a "near Zero" export. In my country its not allowed to export so i have to use that.

**Reply by Danielxyz123** [reply]
Hi, 

valdiressa i also try to get the data from the chint. It's a little bit confusing.
I added a example logfile
This is a line extracted to your hints

3843132012026FCE134C6A2980C9913

Is that correct? How to interpret the raw data?
Thx in advance!

Start Frame | Address Slave/Meter | Function Code | Power data  Reg. | Numb. of Reg. to read | Raw data 4 bytes |    ??
3843             1                                   3                            2012                    02                                 6FCE134C               6A2980C 9913

[Chin_Log.txt](https://github.com/macxq/foxess-ha/files/11956984/Chin_Log.txt)

**Comment by straga** [comment]
Yes.
I have Solax x1 2.0 Gen2 Invertor with Meter for control Export - That works. Couple of seconds if Load is constantly set to 0w.
For the test, I am disconnected Solax and I am connected Foxess S2500 with the meter. All same.
But foxess work can do 0w, always more or less 0. Load around 200w, export/import always -+18 : -+60.
I need to do more tests next week.

**Reply by straga** [reply]
For information. 
Deye/sunsynk- send all info about Modbus.
I was returning foxess to the seller and bought Deye.

**Comment by nathanmarlor** [comment]
Hi all, only just discovered this thread (you guys would've saved me a lot of time 😁) but we've been working on the same thing. I put together a PoC which MITM the WiFi adapter and Fox Cloud and allows you to inject your own modbus read requests etc. You'll need to NAT traffic from the WiFi adapter to the container for this to work.

https://github.com/nathanmarlor/fox-local

Disappointingly however it seems that the WiFi adapter actually has less access than the standard 485 lines - i.e. PV data is inaccessible. Bizarrely it can access the 60000+ registers which is where individual cell voltages/temperatures reside. Because of this I've decided it's a bit of a dead end - my original intent was to have a fully functioning "Fox Local" which simplified monitoring by not needing an external device, wiring into the 485 connector etc. 

Hopefully someone else can look at this with a fresh set of eyes and move it along?! 🙂
